sa/sa.go

```diff
@@ -915,3 +915,210 @@ func (ssa *SQLStorageAuthority) Health(ctx context.Context) error {
 	}
 	return nil
 }
+
+// LeaseCRLShard marks a single crlShards row as leased until the given time.
+// If the request names a specific shard, this function will return an error
+// if that shard is already leased. Otherwise, this function will return the
+// index of the oldest shard for the given issuer.
+func (ssa *SQLStorageAuthority) LeaseCRLShard(ctx context.Context, req *sapb.LeaseCRLShardRequest) (*sapb.LeaseCRLShardResponse, error) {
+	if core.IsAnyNilOrZero(req.Until, req.IssuerNameID) {
+		return nil, errIncompleteRequest
+	}
+	if req.Until.AsTime().Before(ssa.clk.Now()) {
+		return nil, fmt.Errorf("lease timestamp must be in the future, got %q", req.Until.AsTime())
+	}
+
+	if req.MinShardIdx == req.MaxShardIdx {
+		return ssa.leaseSpecificCRLShard(ctx, req)
+	} else {
+		return ssa.leaseOldestCRLShard(ctx, req)
+	}
+}
+
+// leaseOldestCRLShard finds the oldest unleased crl shard for the given issuer
+// and then leases it. Shards within the requested range which have never been
+// leased or are previously-unknown indices are considered older than any other
+// shard. It returns an error if all shards for the issuer are already leased.
+func (ssa *SQLStorageAuthority) leaseOldestCRLShard(ctx context.Context, req *sapb.LeaseCRLShardRequest) (*sapb.LeaseCRLShardResponse, error) {
```

Comment:
I agree when the function is question is a helper that might be called in multiple places, and you want the reader to already know what it does before they see it called, so they don't have to go hunting for it. But in this case, this function (and `leaseSpecificCRLShard` below) is more like "the real thing" while `LeaseCRLShard` is just a wrapper, so I prefer keeping them below so that the whole thing reads more like a single straight-line function.

C.f. `ca.IssuePrecertificate`, where the helper `noteSignError` is defined above, but the heavy-lifting `issuePrecertificateInner` is defined below.